internal/datastore/proxy/relationshipintegrity.go

```diff
@@ -0,0 +1,482 @@
+	ExpiredAt *time.Time
+
+	// Bytes is the raw key material.
+	Bytes []byte
+}
+
+type hmacConfig struct {
+	keyID     string
+	expiredAt *time.Time
+	hasher    hash.Hash
+	lock      sync.Mutex
+}
+
+var (
+	versionByte = byte(0x01)
+	hashLength  = 16
+)
+
+// NewRelationshipIntegrityProxy creates a new datastore proxy that ensures the integrity of
+// relationships by using HMACs to sign the data. The current key is used to sign new data,
+// and the expired keys are used to verify old data, if any.
+func NewRelationshipIntegrityProxy(ds datastore.Datastore, currentKey KeyConfig, expiredKeys []KeyConfig) (datastore.Datastore, error) {
+	// Ensure the datastore supports integrity.
+	unwrapped := ds
+	if wds, ok := unwrapped.(datastore.UnwrappableDatastore); ok {
+		unwrapped = wds.Unwrap()
+	}
+
+	id, ok := unwrapped.(datastore.IntegrityDatastore)
+	if !ok {
+		return nil, fmt.Errorf("datastore of kind %T does not implement relationship integrity", ds)
+	}
+
+	if !id.SupportsIntegrity() {
+		return nil, spiceerrors.MustBugf("datastore does not support relationship integrity")
+	}
+
+	if len(currentKey.Bytes) == 0 {
+		return nil, fmt.Errorf("contents of the current key file cannot be empty")
+	}
+
+	if len(currentKey.ID) == 0 {
+		return nil, fmt.Errorf("current key ID cannot be empty")
+	}
+
+	hasher, err := hasherForKey(currentKey.Bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	currentKeyHMAC := &hmacConfig{
+		keyID:     currentKey.ID,
+		expiredAt: currentKey.ExpiredAt,
+		hasher:    hasher,
+		lock:      sync.Mutex{},
+	}
+
+	if currentKey.ExpiredAt != nil {
+		return nil, spiceerrors.MustBugf("current key cannot have an expiration")
+	}
+
+	keysByID := make(map[string]*hmacConfig, len(expiredKeys)+1)
+	keysByID[currentKey.ID] = currentKeyHMAC
+
+	expiredKeyIDs := make([]string, 0, len(expiredKeys))
+	for _, key := range expiredKeys {
+		if len(key.Bytes) == 0 {
+			return nil, fmt.Errorf("expired key cannot be empty")
+		}
+
+		if len(key.ID) == 0 {
+			return nil, fmt.Errorf("expired key ID cannot be empty")
+		}
+
+		if key.ExpiredAt == nil {
+			return nil, fmt.Errorf("expired key missing expiration time")
+		}
+
+		if _, ok := keysByID[key.ID]; ok {
+			return nil, fmt.Errorf("found duplicate key ID: %s", key.ID)
+		}
+
+		hasher, err := hasherForKey(key.Bytes)
+		if err != nil {
+			return nil, err
+		}
+
+		keysByID[key.ID] = &hmacConfig{
+			keyID:     key.ID,
+			expiredAt: key.ExpiredAt,
+			hasher:    hasher,
+		}
+
+		expiredKeyIDs = append(expiredKeyIDs, key.ID)
+	}
+
+	log.Debug().
+		Str("current_key_id", currentKey.ID).
+		Strs("expired_key_ids", expiredKeyIDs).
+		Msgf("created relationship integrity proxy")
+
+	return &relationshipIntegrityProxy{
+		ds:         ds,
+		primaryKey: currentKeyHMAC,
+		keysByID:   keysByID,
+	}, nil
+}
+
+func hasherForKey(key []byte) (hash.Hash, error) {
+	return hmac.New(sha256.New, key), nil
+}
+
+type relationshipIntegrityProxy struct {
+	ds         datastore.Datastore
```

Comment:
if you just embed  `datastore.Datastore` instead of naming it, you can remove most of the methods in this file since they'll just pass through

Reply:
Exactly what I didn't want to be possible - if we add another read method down the road and we just embed, this won't fail to compile. I want it to do so to ensure that we don't accidentally forget to add validation to any new read method added